main/ts_serial.c

```diff
@@ -287,65 +287,77 @@ int ts_serial_scan_device_info(TSDevice *device)
     return 0;
 }
 
-int ts_serial_ota(int chip_type)
+static int finish_ota(FILE *f, int ret)
+{
+    if (f != NULL) {
+        fclose(f);
+    }
+    // maybe this should be executed only for ret == ESP_OK
+    stm32bl_reset_device();
+    uart_set_parity(uart_num, UART_PARITY_DISABLE);
+    xSemaphoreGive(uart_lock);
+    return ret;
+}
+
+int ts_serial_ota(int flash_size, int page_size)
 {
     int ret = ESP_OK;
+    uint16_t pages = flash_size / page_size;
 
     ts_serial_request("!exec/bootloader-stm\n", 100);
     ts_serial_response_clear();
 
     // prevent further UART access in RX thread
-    if (xSemaphoreTake(uart_lock, pdMS_TO_TICKS(100)) == pdFALSE) {
-        // this is bad and should not happen, as we already started the bootloader now
-        ESP_LOGE(TAG, "Could not take semaphore uart_lock");
-        return ESP_FAIL;
+    int retry = 0;
+    while (xSemaphoreTake(uart_lock, pdMS_TO_TICKS(100)) != pdTRUE ) {
+        vTaskDelay(pdMS_TO_TICKS(100));
+        retry++;
+        if (retry > OTA_MAX_RETRY) {
+             // this is bad and should not happen, as we already started the bootloader now
+            ESP_LOGE(TAG, "Could not take semaphore uart_lock");
+            return ESP_FAIL;
+        }
```

Comment:
Why not just increase the wait time for the Semaphore by a factor of `OTA_MAX_RETRY` instead? Do we really need a loop with multiple tries?